apis/apps/v1alpha1/opsrequest_types.go

```diff
@@ -311,12 +311,33 @@ type Upgrade struct {
 	//
 	// This field is deprecated since v0.9 because ClusterVersion is deprecated.
 	//
-	// +kubebuilder:validation:Required
 	// +kubebuilder:deprecatedversion:warning="This field has been deprecated since 0.9.0"
-	ClusterVersionRef string `json:"clusterVersionRef"`
+	ClusterVersionRef string `json:"clusterVersionRef,omitempty"`
+
+	// Lists components to be upgrade based on desired ComponentDefinition or ServiceVersion.
+	// +patchMergeKey=componentName
+	// +patchStrategy=merge,retainKeys
+	// +listType=map
+	// +listMapKey=componentName
+	// +kubebuilder:validation:MaxItems=1024
+	Components []UpgradeComponent `json:"components,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"componentName"`
 }
 
-// VerticalScaling refers to the process of adjusting the compute resources (e.g., CPU, memory) allocated to a Component.
+// +kubebuilder:validation:XValidation:rule="has(self.componentDefinitionName) || has(self.serviceVersion)",message="at least one componentDefinitionName or serviceVersion"
+
+type UpgradeComponent struct {
+	// Specifies the name of the Component.
+	ComponentOps `json:",inline"`
+
+	// Specifies the name of the ComponentDefinition.
```

Comment:
Should allows componentDefinitionName to be optional.
Case 1: componentDefinitionName is empty + non-empty serviceVersion: keep componentDefinitionName unchanged and use new release.
Case 2: non-empty componentDefinitionName + empty serviceVersion: upgrade compdef, use the latest available release.
Case 3: non-empty componentDefinitionName + non-empty serviceVersion: upgrade compdef, and use the specified release.

Reply:
The reasonable combinations should be:
1. (comp-def, service-ver) - upgrade to the specified service version and component definition, the user takes the responsibility to ensure that they are compatible.
2. (nil, service-ver) - upgrade to the specified service version, let the operator to choose the latest compatible component definition.
3. (comp-def, nil) - upgrade to the specified component definition, let the operator to choose the latest compatible service version.
4.  (nil, nil) - upgrade to the latest service version and component definition, the operator will ensure the compatibility between the selected versions.